---
title: "Read_normalization"
output: html_document
date: "2024-07-25"
---

#Load your input file from idxstats.sh script

```{r data}

data <- read.table("reads_per_chromosome.txt", header = TRUE, sep = "\t") #the input file was generated by runnning samtools idxstats on BAM files (idxstats.sh script). 

```

#Generate coverage plot

```{r generate depth coverage plot}

long_data <- data %>%
  pivot_longer(
    cols = starts_with("chr"),  #your chromosome columns should start with 'chr'
    names_to = "Chromosome",
    values_to = "Depth"
  ) %>%
  mutate(Chromosome = factor(Chromosome, levels = paste0("chr", 1:11)))

depth_plot <- ggplot(long_data, aes(x = Chromosome, y = Depth, group = Individual, color = as.factor(Individual))) +
  geom_line() +
  labs(
    x = "Chromosome",
    y = "Number of reads (x1000)"
  ) +
  scale_y_continuous(labels = function(x) x / 1000) + 
  theme_light(base_size = 14) + 
  theme(
    legend.position = "none",
    panel.background = element_rect(fill = "white"),  
    panel.grid.major = element_blank(),  
    panel.grid.minor = element_blank(),  
    panel.border = element_blank(), 
    axis.line = element_line(color = "black") 
  )

ggsave("/Users/your_user_name/plot_depth.png", plot = depth_plot, device = "png", width = 10, height = 6, dpi = 300)



```

#Function to calculate alpha

```{r alpha-function}

library(tidyverse)

#function to calculate alpha values
calculate_alpha <- function(data) {
  data_numeric <- data[, -1]
  N <- sum(data_numeric)
  ni <- rowSums(data_numeric)
  nj <- colSums(data_numeric)
  
  alpha_matrix <- matrix(NA, nrow=nrow(data_numeric), ncol=ncol(data_numeric))
  rownames(alpha_matrix) <- data[, 1]
  colnames(alpha_matrix) <- names(nj)
  
  for (i in 1:nrow(data_numeric)) {
    for (j in 1:ncol(data_numeric)) {
      nij <- data_numeric[i, j]
      alpha_matrix[i, j] <- (nij * 1.0 * N) / (ni[i] * 1.0 * nj[j])
    }
  }
  alpha_df <- as.data.frame(alpha_matrix)
  alpha_melted <- alpha_df %>% 
    rownames_to_column(var = "sample") %>% 
    pivot_longer(cols = -sample, names_to = "CHROM", values_to = "alpha")
  return(alpha_melted)
}

# Calculate alpha values 
alpha_df <- calculate_alpha(data)


```


#Below is the specific code for our two datasets (this study and De La Torre et al, 2021 dataset)


```{r Calculate alpha for two datasets}


library(magrittr)
library(dplyr)
library(tidyr)
library(tibble)
library(ggplot2)
library(gridExtra)


get_legend <- function(myggplot) {
  tmp <- ggplot_gtable(ggplot_build(myggplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)
}

# Read the datasets
data <- read.table("/Users/sashanikolaeva/reads_per_chromosome_no_chloroplast.txt", header = TRUE, sep = "\t")

alpha_df_data <- calculate_alpha(data)
alpha_df_data$source <- "Chlor"

data_tissue_culture <- read.table("/Users/sashanikolaeva/reads_per_chromosome_DeLaTorre.txt", header = TRUE)
alpha_df_tissue_culture <- calculate_alpha(data_tissue_culture)
alpha_df_tissue_culture$source <- "tissue_culture"

# Identify points with gain or loss of chromosomes
alpha_df_data <- alpha_df_data %>%
  mutate(point_color = case_when(
    alpha > 1.05 ~ "Gain",
    alpha < 0.95 ~ "Loss",
    TRUE ~ "Neutral"
  ))

data_tissue_culture <- data_tissue_culture %>%
  mutate(point_color = case_when(
    alpha > 1.05 ~ "Gain",
    alpha < 0.95 ~ "Loss",
    TRUE ~ "Neutral"
  ))

# Ensure chromosome numbers are treated as a factor with levels in the correct order
alpha_df_data$CHROM <- factor(alpha_df_data$CHROM, levels = paste("chr", 1:11, sep = ""))
alpha_df_tissue_culture$CHROM <- factor(alpha_df_tissue_culture$CHROM, levels = paste("chr", 1:11, sep = ""))

# Plot for this study dataset
plot_data <- ggplot(alpha_df_data, aes(x = CHROM, y = alpha, group = sample)) +
  geom_line(color = "grey", linewidth = 0.5, alpha = 0.5) +
  geom_point(aes(color = point_color), size = 1) +
  geom_hline(yintercept = 1.05, color = "red", linetype = "dashed") +
  geom_hline(yintercept = 0.95, color = "blue", linetype = "dashed") +
  ylab(expression(alpha * "-value")) +
  xlab("Chromosome") +
  annotate("text", x = 6, y = 0.92, label = "This study", hjust = 0.5, vjust = 1.5, size = 5, fontface = "bold") +
  theme_light(base_size = 14) +
  theme(
    panel.background = element_rect(fill = "white"),
    panel.grid.major = element_blank(),  # Removes major grid lines
    panel.grid.minor = element_blank(),  # Removes minor grid lines
    panel.border = element_blank(),  # Removes border around the plot
    axis.line = element_line(color = "black"),  # Adds a line along the axes for clarity
    legend.position = "none"  # Removes legend
  ) +
  scale_color_manual(values = c("Gain" = "red", "Loss" = "blue", "Neutral" = "grey"))

# Plot for tissue culture dataset
plot_tissue_culture <- ggplot(alpha_df_tissue_culture, aes(x = CHROM, y = alpha, group = sample)) +
  geom_line(color = "grey", linewidth = 0.5, alpha = 0.5) +
  geom_point(aes(color = point_color), size = 1) +
  geom_hline(yintercept = 1.05, color = "red", linetype = "dashed") +
  geom_hline(yintercept = 0.95, color = "blue", linetype = "dashed") +
  ylab(expression(alpha * "-value")) +
  xlab("Chromosome") +
  annotate("text", x = 6, y = 0.92, label = "De La Torre et al, 2021", hjust = 0.5, vjust = 1.5, size = 5, fontface = "bold") +
  theme_light(base_size = 14) +
  theme(
    panel.background = element_rect(fill = "white"),
    panel.grid.major = element_blank(),  # Removes major grid lines
    panel.grid.minor = element_blank(),  # Removes minor grid lines
    panel.border = element_blank(),  # Removes border around the plot
    axis.line = element_line(color = "black")  # Adds a line along the axes for clarity
  ) +
  scale_color_manual(values = c("Gain" = "red", "Loss" = "blue", "Neutral" = "grey")) +
  labs(color = "Chromosome change")  # Title for the legend

# Create a common legend
common_legend <- get_legend(plot_tissue_culture)

# Arrange the plots side by side with the common legend
combined_plot <- grid.arrange(
  arrangeGrob(plot_data, plot_tissue_culture + theme(legend.position = "none"), ncol = 2),
  common_legend, ncol = 2, widths = c(10, 1)
)

# Save the combined plot to a file
ggsave("/Users/sashanikolaeva/combined_plot_alpha_side_by_side_with_legend.png", plot = combined_plot, device = "png", width = 14, height = 6, dpi = 300)


# Identify samples with alpha values above 1.05 or below 0.95
highlight_samples_data <- alpha_df_data %>%
  group_by(sample) %>%
  summarize(has_extreme_values = any(alpha > 1.05 | alpha < 0.95)) %>%
  filter(has_extreme_values) %>%
  pull(sample)

highlight_samples_tissue_culture<- alpha_df_tissue_culture %>%
  group_by(sample) %>%
  summarize(has_extreme_values = any(alpha > 1.05 | alpha < 0.95)) %>%
  filter(has_extreme_values) %>%
  pull(sample)

# Print the number of samples with deviations
num_blue_samples <- length(highlight_samples_data)
num_red_samples <- length(highlight_samples_tissue_culture)

cat("Number of samples (Loss):", num_blue_samples, "\n")
cat("Number of samples(Gain):", num_red_samples, "\n")



```


```{r alpha by PCR group}

library(magrittr)
library(dplyr)
library(tidyr)
library(tibble)
library(ggplot2)
library(gridExtra)
library(stringr)

# Extracting PoolNumber from the 'Individual' column
data$PoolNumber <- as.numeric(str_extract(data$Individual, "(?<=RedwoodCap_)(\\d+)(?=_)")) 

# Assigning the RCRGroup values based on the number of PCR cycles
data$RCRGroup <- case_when(
  data$PoolNumber %in% c(14, 17, 18, 22) ~ 1,
  data$PoolNumber %in% c(13, 15, 16, 19, 20, 21, 23, 24) ~ 2,
  data$PoolNumber %in% c(26, 27, 29, 31, 32, 33, 35, 36) ~ 3,
  data$PoolNumber %in% c(1, 2, 3, 4, 5, 6, 7, 8) ~ 4,
  data$PoolNumber %in% c(9, 10, 11, 12, 25, 28, 30, 34) ~ 5,
  TRUE ~ NA_integer_
)

# Define columns to keep (excluding RCRGroup and PoolNumber)
columns_to_keep <- setdiff(colnames(data), c("RCRGroup", "PoolNumber"))

# Splitting data into separate dataframes based on RCRGroup
data_df_1 <- subset(data, RCRGroup == 1)
data_df_2 <- subset(data, RCRGroup == 2)
data_df_3 <- subset(data, RCRGroup == 3)
data_df_4 <- subset(data, RCRGroup == 4)
data_df_5 <- subset(data, RCRGroup == 5)

# Remove the 'RCRGroup' and 'PoolNumber' columns
data_df_1 <- data_df_1[, !(names(data_df_1) %in% c("PoolNumber", "RCRGroup"))]
data_df_2 <- data_df_2[, !(names(data_df_2) %in% c("PoolNumber", "RCRGroup"))]
data_df_3 <- data_df_3[, !(names(data_df_3) %in% c("PoolNumber", "RCRGroup"))]
data_df_4 <- data_df_4[, !(names(data_df_4) %in% c("PoolNumber", "RCRGroup"))]
data_df_5 <- data_df_5[, !(names(data_df_5) %in% c("PoolNumber", "RCRGroup"))]

alpha_df1 <- calculate_alpha(data_df_1)
alpha_df2 <- calculate_alpha(data_df_2)
alpha_df3 <- calculate_alpha(data_df_3)
alpha_df4 <- calculate_alpha(data_df_4)
alpha_df5 <- calculate_alpha(data_df_5)

# Extracting PoolNumber for each alpha_df
alpha_df1$PoolNumber <- as.numeric(str_extract(alpha_df1$sample, "(?<=RedwoodCap_)(\\d+)(?=_)"))
alpha_df2$PoolNumber <- as.numeric(str_extract(alpha_df2$sample, "(?<=RedwoodCap_)(\\d+)(?=_)"))
alpha_df3$PoolNumber <- as.numeric(str_extract(alpha_df3$sample, "(?<=RedwoodCap_)(\\d+)(?=_)"))
alpha_df4$PoolNumber <- as.numeric(str_extract(alpha_df4$sample, "(?<=RedwoodCap_)(\\d+)(?=_)"))
alpha_df5$PoolNumber <- as.numeric(str_extract(alpha_df5$sample, "(?<=RedwoodCap_)(\\d+)(?=_)"))

# Function to count samples with deviations
count_deviations <- function(df) {
  sum(df$alpha > 1.05 | df$alpha < 0.95)
}

# Count deviations for each group
deviations_1 <- count_deviations(alpha_df1)
deviations_2 <- count_deviations(alpha_df2)
deviations_3 <- count_deviations(alpha_df3)
deviations_4 <- count_deviations(alpha_df4)
deviations_5 <- count_deviations(alpha_df5)

# Calculate total number of samples with deviations
total_deviations <- deviations_1 + deviations_2 + deviations_3 + deviations_4 + deviations_5

cat("Number of samples with deviations in PCR Group 1:", deviations_1, "\n")
cat("Number of samples with deviations in PCR Group 2:", deviations_2, "\n")
cat("Number of samples with deviations in PCR Group 3:", deviations_3, "\n")
cat("Number of samples with deviations in PCR Group 4:", deviations_4, "\n")
cat("Number of samples with deviations in PCR Group 5:", deviations_5, "\n")
cat("Total number of samples with deviations:", total_deviations, "\n")

# Function to create plots 
create_plot <- function(df, group_title) {
  df <- df %>%
    mutate(point_color = case_when(
      alpha > 1.05 ~ "Gain",
      alpha < 0.95 ~ "Loss",
      TRUE ~ "Neutral"
    ))
  
  ggplot(df, aes(x = factor(CHROM, levels = paste0("chr", 1:11)), y = alpha, group = sample)) +
    geom_line(color = "grey", linewidth = 0.5, alpha = 0.5) +
    geom_point(aes(color = point_color), size = 1) +
    geom_hline(yintercept = 1.05, color = "red", linetype = "dashed") +
    geom_hline(yintercept = 0.95, color = "blue", linetype = "dashed") +
    ylab(expression(alpha)) + xlab("Chromosome") +
    theme_classic() +
    theme(axis.title = element_text(size = 17), axis.text = element_text(size = 15)) + 
    scale_color_manual(values = c("Gain" = "red", "Loss" = "blue", "Neutral" = "grey"), guide = "none") +
    ggtitle(group_title)
}

# Plotting for each dataframe
plot_group1 <- create_plot(alpha_df1, "PCR Group 1")
plot_group2 <- create_plot(alpha_df2, "PCR Group 2")
plot_group3 <- create_plot(alpha_df3, "PCR Group 3")
plot_group4 <- create_plot(alpha_df4, "PCR Group 4")
plot_group5 <- create_plot(alpha_df5, "PCR Group 5")

plot_group1
plot_group2
plot_group3
plot_group4
plot_group5





```

